lib/ui/painting/multi_frame_codec.cc

```diff
@@ -146,9 +146,14 @@ MultiFrameCodec::State::GetNextFrameImage(
     return std::make_pair(nullptr, decode_error);
   }
 
-  // Hold onto this if we need it to decode future frames.
-  if (frameInfo.disposal_method == SkCodecAnimation::DisposalMethod::kKeep ||
-      lastRequiredFrame_) {
+  // Hold onto this if we need it to decode future frames except that current
+  // frame's disposal_method is kRestorePrevious.
+  const bool dont_keep_current_frame =
+      frameInfo.disposal_method ==
+      SkCodecAnimation::DisposalMethod::kRestorePrevious;
+  if (!dont_keep_current_frame &&
+      (frameInfo.disposal_method == SkCodecAnimation::DisposalMethod::kKeep ||
+       lastRequiredFrame_)) {
```

Comment:
Ah, yeah, this simplifies to `kKeep || (lastRequiredFrame_ && !kRestorePrevious)`, which makes sense: When the disposal OP is `kRestorePrevious`, we will continue retaining the last frame stored, if any. Great!

I suggest refactoring this logic & surrounding comments to something like:
```cpp
  const bool keep_current_frame = frameInfo.disposal_method == SkCodecAnimation::DisposalMethod::kKeep;
  const bool restore_previous_frame = frameInfo.disposal_method == SkCodecAnimation::DisposalMethod::kRestorePrevious;
  const bool previous_frame_available = lastRequiredFrame_.has_value();

  // Store the current frame in `lastRequiredFrame_` if the frame's disposal
  // method indicates we should do so.
  // * When the disposal method is "Keep", the stored frame should always be
  //   overwritten with the new frame we just crafted.
  // * When the disposal method is "RestorePrevious", the previously stored
  //   frame should be retained and used as the backdrop for the next frame
  //   again. If there isn't already a stored frame, that means we haven't
  //   rendered any frames yet! When this happens, we just fall back to "Keep"
  //   behavior and store the current frame as the backdrop of the next frame.

  if (keep_current_frame || (previous_frame_available && !restore_previous_frame)) {
    // Replace the stored frame. The `lastRequiredFrame_` will get used as the starting
    // backdrop for the next frame.
    lastRequiredFrame_ = std::make_unique<SkBitmap>(bitmap);
    lastRequiredFrameIndex_ = nextFrameIndex_;
  }
```
Which should make this whole state machine much easier to understand the next time someone has to modify it. 😅

Reply:
The refactoring is great, it seems much easier to understand. 

[nice.gif]